Fix Rename-Item -Path with wildcard char

kwkam opened the pull request:
## PR Summary

_*Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:_
Unescape non-literal, non-glob path in ProccessRecord and
set the context.SuppressWildcardExpansion in RenameItem.
This solve the issue where Rename-Item complains -Path does not exist
when both -Path and CWD contains special characters.

_*System.Management.Automation/namespaces/LocationGlobber.cs:_
Do not escape CWD when LiteralPath is used.
This causes issues when both -LiteralPath and CWD contains special characters. For example,
```PowerShell
Rename-Item -LiteralPath [test]file.txt [test]file2.txt
```
will complain that ```[test]file.txt``` does not exist when CWD contains special characters.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

kwkam committed:
[Feature] Fix Rename-Item -Path with wildcard char

Set the context.SuppressWildcardExpansion in MoveItem instead of
escaping the path every time. Also solve the issue where Move-Item
complains -Path wildcard pattern is not valid when -Path contains
special characters that forms an invalid pattern.

kwkam committed:
[Feature] Fix FunctionProvider test

kwkam committed:
fixup! [Feature] Fix FunctionProvider test

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,99 @@ protected override bool ProviderSupportsShouldProcess
         /// </summary>
         protected override void ProcessRecord()
+        {
+            do
```
Why do we use cycle here? We have single input string.
I would try to make things easier - just try to rename in `try-catch`.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/resources/NavigationResources.resx:
```diff
@@ -168,6 +168,9 @@
     <value>Cannot move item because the item at '{0}' is in use.</value>
   </data>
+  <data name="RenameMultipleItemError" xml:space="preserve">
+      <value>Cannot rename item because the path '{0}' resolved to multiple items. Only one item can be renamed at a time.</value>
```
It makes no sense to tell the user that we have tried to resolve wildcards - in fact, we consider the Path parameter as literal.

kwkam replied:
I wonder if we would add support for renaming multiple items like the `ren` command in cmd?
eg. ```Rename-Item -Path '*.txt' -NewExtension '.bak'```
or, ```Rename-Item -Path '*.txt' -Format 'renamed_{0}'```

iSazonov replied:
I believe the enhancement is out of the PR and we should discuss it in a new tracking Issue (please open).

iSazonov reviewed (commented):
Leave a comment

kwkam committed:
LocationGlobber: Fix issue when CWD contains wildcard char

Do not escape CWD when LiteralPath is used.
This causes issues when both -LiteralPath and CWD contains special characters.

kwkam committed:
[Feature] Treat Path as literal

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,93 @@ protected override bool ProviderSupportsShouldProcess
+                }
+            }
+            finally {
+            }
```
We can remove the empty finally block.

kwkam committed:
[Feature] Remove unnecessary finally

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,91 @@ protected override bool ProviderSupportsShouldProcess
         protected override void ProcessRecord()
+        {
+            try
+            {
```
Seem all used methods use `try-catch` and we can remove the `try`.

kwkam committed:
Revert "[Feature] Remove unnecessary finally"

This reverts commit be59b3d05f189ec569ff55b99bce752e10105f5c.

kwkam committed:
[Feature] Remove try-finally

iSazonov commented:
@kwkam Please rebase to fix CI Travis failure.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,88 @@ protected override bool ProviderSupportsShouldProcess
 
+        private Collection<PathInfo> GetResolvedPaths(string path)
+        {
+            Collection<PathInfo> results = new Collection<PathInfo>();
```
It is extra allocation. I think we should remove this and check null in calling code.

kwkam replied:
Updated test.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,88 @@ protected override bool ProviderSupportsShouldProcess
+                    break;
+
+                default:
+                    RenameItem(Path, literalPath: true);
```
Is this really working?

iSazonov replied:
I'd expect that we unescape path, check existance (or try rename), if fail then resolve globs (write an error if cpunt > 1) and try rename again.

kwkam replied:
I believe -Path should try to resolve globs first, since it's the expected behaviour, and users can escape the wildcard character, for example
```PowerShell
Rename-Item -Path '`[A-B`]' -NewName 'A-B'
```
will prevent globbing.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3603,22 +3603,88 @@ protected override bool ProviderSupportsShouldProcess
         /// </summary>
         protected override void ProcessRecord()
+        {
+            if (base.SuppressWildcardExpansion)
```
Do we really need `base` prefix?

kwkam replied:
Removed.

iSazonov reviewed (commented):
Please look tests - I don't see test related the PR.

kwkam committed:
Move Remove-Item test to #7404

kwkam committed:
Fix commented items

kwkam committed:
[Feature] Add test

kwkam committed:
fixup! Fix commented items

kwkam committed:
fixup! [Feature] Add test

kwkam commented:
@iSazonov Should I run `git rebase master`?

kwkam committed:
fixup! fixup! [Feature] Add test

iSazonov commented:
Why do you need to rebase?

kwkam commented:
@iSazonov https://github.com/PowerShell/PowerShell/pull/7398#issuecomment-409798119

iSazonov commented:
@kwkam Ok, I understand. You can rebase to temporary skip broken tests.

kwkam commented:
@iSazonov I see. I would not rebase if it is not necessary.

kwkam committed:
[Feature] Add test for -LiteralPath

iSazonov commented:
@SteveL-MSFT do we have interest in the change?

SteveL-MSFT commented:
@iSazonov yes, interested in this change

iSazonov commented:
@kwkam Please rebase to pass CIs.

kwkam committed:
Merge branch 'master' into wc-renameitem

kwkam commented:
@iSazonov OK, rebased to latest master.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3519,22 +3519,87 @@ protected override bool ProviderSupportsShouldProcess
+        private void RenameItem(string path, bool literalPath = false)
         {
             CmdletProviderContext currentContext = CmdletProviderContext;
+            currentContext.SuppressWildcardExpansion = literalPath;
```
seems like `literalPath` doesn't need to be passed around as you can just use `SuppressWildcardExpansion` from base?

kwkam replied:
I believe it would be unable to resolve `-Path` as literal path if we use the `base.SuppressWildcardExpansion` here.

SteveL-MSFT reviewed (approved):
LGTM

adityapatwardhan commented:
@kwkam Thank you for your contribution!
